src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -0,0 +1,345 @@
+                    left._referenceHandler == right._referenceHandler &&
+                    left._encoder == right._encoder &&
+                    left._defaultIgnoreCondition == right._defaultIgnoreCondition &&
+                    left._numberHandling == right._numberHandling &&
+                    left._unknownTypeHandling == right._unknownTypeHandling &&
+                    left._defaultBufferSize == right._defaultBufferSize &&
+                    left._maxDepth == right._maxDepth &&
+                    left._allowTrailingCommas == right._allowTrailingCommas &&
+                    left._ignoreNullValues == right._ignoreNullValues &&
+                    left._ignoreReadOnlyProperties == right._ignoreReadOnlyProperties &&
+                    left._ignoreReadonlyFields == right._ignoreReadonlyFields &&
+                    left._includeFields == right._includeFields &&
+                    left._propertyNameCaseInsensitive == right._propertyNameCaseInsensitive &&
+                    left._writeIndented == right._writeIndented &&
+                    left._serializerContext == right._serializerContext &&
+                    CompareConverters(left.Converters, right.Converters);
+
+                static bool CompareConverters(IList<JsonConverter> left, IList<JsonConverter> right)
+                {
+                    int n;
+                    if ((n = left.Count) != right.Count)
+                    {
+                        return false;
+                    }
+
+                    for (int i = 0; i < n; i++)
+                    {
+                        if (left[i] != right[i])
+                        {
+                            return false;
+                        }
+                    }
+
+                    return true;
+                }
+            }
+
+            public int GetHashCode(JsonSerializerOptions options)
+            {
+                HashCode hc = default;
+
+                hc.Add(options._dictionaryKeyPolicy);
+                hc.Add(options._jsonPropertyNamingPolicy);
+                hc.Add(options._readCommentHandling);
+                hc.Add(options._referenceHandler);
+                hc.Add(options._encoder);
+                hc.Add(options._defaultIgnoreCondition);
+                hc.Add(options._numberHandling);
+                hc.Add(options._unknownTypeHandling);
+                hc.Add(options._defaultBufferSize);
+                hc.Add(options._maxDepth);
+                hc.Add(options._allowTrailingCommas);
+                hc.Add(options._ignoreNullValues);
+                hc.Add(options._ignoreReadOnlyProperties);
+                hc.Add(options._ignoreReadonlyFields);
+                hc.Add(options._includeFields);
+                hc.Add(options._propertyNameCaseInsensitive);
+                hc.Add(options._writeIndented);
+                hc.Add(options._serializerContext);
+
+                foreach (JsonConverter converter in options.Converters)
```

Comment:
Iterating over an IList allocates an enumerator, which means this GetHashCode implementation is allocating, which you probably want to avoid.  It's also incurring multiple interface dispatch for MoveNext/Current.  Consider instead iterating via the IList indexer, iterating over the underlying `List<T>` if you can get access to it, or even leaving the converters out of the hash code entirely.